### PR TITLE
Optionally build with source dependencies for akka

### PR DESCRIPTION
**Pull request by jrudolph**

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -1,14 +1,21 @@
 import com.typesafe.sbt.SbtMultiJvm.MultiJvmKeys.MultiJvm
 import akka._
 
+val akkaRef = {
+  // as a little hacky side effect also disable aggregation of samples
+  sys.props += "akka.build.aggregateSamples" -> "false"
+
+  uri("git://github.com/akka/akka.git#release-2.4")
+}
```
Comment (ktoso) [inline build.sbt]: Nice!

**Comment by 2m** [comment]
Great stuff. This reminded me of https://github.com/sbt/sbt/pull/2844

We could implement it similarly: if there is `-Dakka.path` set, depend on sources.

**Comment by ktoso** [comment]
Yeah, we'd need it like that I believe.

**Comment by jrudolph** [comment]
> We could implement it similarly: if there is -Dakka.path set, depend on sources.

I now called the environment variable `akka.sources` which needs to be a URI either to a git repo or a local file. I removed the redundant boolean setting to make the switch.

I already fixed two tests that wouldn't compile against current akka master anymore.

File: project/Dependencies.scala [hunk 2]
```diff
@@ -113,39 +107,27 @@ object Dependencies {
...
+  lazy val http2 = l ++= Seq(hpack, alpnApi)
 
   lazy val httpTestkit = l ++= Seq(
-    akkaStreamTestkit,
     Test.junit, Test.junitIntf, Compile.junit % "provided",
     Test.scalatest.value.copy(configurations = Some("provided; test"))
   )
 
-  lazy val httpTests = l ++= Seq(Test.akkaMmltinodeTestKit, Test.junit, Test.scalatest.value, Test.junitIntf)
+  lazy val httpTests = l ++= Seq(Test.junit, Test.scalatest.value, Test.junitIntf)
```
Comment (ktoso) [inline project/Dependencies.scala]: multinode is used for the `perfspec` though AFAIR?

File: project/AkkaDependency.scala [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka
+
+import sbt._
+import Keys._
+
+object AkkaDependency {
+  // Needs to be a URI like git://github.com/akka/akka.git#master or file:///xyz/akka
+  val akkaSourceDependencyUri = sys.props.getOrElse("akka.sources", "")
```
Comment (ktoso) [inline project/AkkaDependency.scala]: excellent 👍

File: akka-http-core/src/test/scala/akka/http/impl/engine/ws/WSClientAutobahnTest.scala [hunk 4]
```diff
@@ -167,28 +164,8 @@ object WSClientAutobahnTest extends App {
...
-          override def onUpstreamFailure(cause: Throwable, ctx: Context[T]): TerminationDirective = {
-            p.failure(cause)
-            super.onUpstreamFailure(cause, ctx)
-          }
-        }
-
-      (stage, p.future)
-    }
+  def completionSignal[T]: Flow[T, T, Future[Done]] =
+    Flow[T].watchTermination()((_, res) ⇒ res)
```
Comment (ktoso) [inline akka-http-core/src/test/scala/akka/http/impl/engine/ws/WSClientAutobahnTest.scala]: 👍

File: build.sbt [hunk 5]
```diff
@@ -88,6 +93,7 @@ lazy val httpTests = project("akka-http-tests")
   .enablePlugins(MultiNode)
   .disablePlugins(MimaPlugin) // this is only tests
   .configs(MultiJvm)
+  .addAkkaModuleDependency("akka-multi-node-testkit", "test")
```
Comment (ktoso) [inline build.sbt]: oh I see, okey

**Comment by ktoso** [comment]
LGTM, nice one!

**Comment by jrudolph** [comment]
Thanks for merging. I'll setup jobs to run the builds nightly.

**Pull request merged**